airflow/models/dagrun.py

```diff
@@ -552,6 +552,9 @@ def is_effective_leaf(task):
             return not task.is_teardown or task.on_failure_fail_dagrun
 
         leaf_task_ids = {x.task_id for x in dag.tasks if is_effective_leaf(x)}
+        if not leaf_task_ids:
+            # can happen if dag is exclusively teardown tasks
+            leaf_task_ids = {x.task_id for x in dag.tasks if not x.downstream_list}
```

Comment:
Why do we need a non-empty list of leaves? Does it affect the DAG run state or serve a different purpose?

Reply:
Yes you need something to infer dag run state

Reply:
So when the teardown task fails, we will fail the dag run even if `on_failure_fail_dagrun` is not `True`? I wonder if we can just consider the empty leaves list as all success without looking for the teardown tasks state :thinking:

Reply:
Yeah it's an odd case. Very much nonsensical edge case that doesn't really make sense. If there's no no teardown tasks then there's no point in ignoring the teardowns. There's no point in even having tasks marked as teardowns in that case. So this seemed like a sensible fallback but we can discuss